Add how-to doc on refresh tokens (#778)
jupyterhub/oauthenticator: docs/source/how-to/refresh.md

# Refreshing user authentication

> JupyterHub has a mechanism called [`refresh_user`](inv:jupyterhub:py:method#jupyterhub.auth.Authenticator.refresh_user) that is meant to 'refresh' information from the Authentication provider periodically.

Reviewer: ```suggestion
JupyterHub has a mechanism called [`refresh_user`](inv:jupyterhub:py:method#jupyterhub.auth.Authenticator.refresh_user) that is meant to _refresh_ information from the Authentication provider periodically.
```
I think italics should look better when rendered:
![image](https://github.com/user-attachments/assets/0e8420cf-51ac-4657-a65c-fcb8ef4fc1a5)

  then the user must login again before they are able to take actions in JupyterHub
  because at this point their authorization state is unknown and could no longer be valid.

There is also an option [Authenticator.refresh_pre_spawn](inv:jupyterhub:py:attribute#jupyterhub.auth.Authenticator.refresh_pre_spawn) which can be enabled:

```python
c.Authenticator.refresh_pre_spawn = True
```

to ensure auth is up-to-date before launching a server.
This is most useful when the server is being passed an access token
because it ensures the token is valid when the server starts.

## Refreshing tokens from user sessions

If your user sessions use access tokens from your oauth provider and those tokens may expire during user sessions,
you can rely on this mechanism to get fresh access tokens from JupyterHub.

The first step is to grant the _server_ token access to read auth state for its owner.
Users do not have permission to read their own auth state by default,
but `auth_state` is where the `access_token` is stored.
We need to grant the `admin:auth_state!user` scope to both the `user` and `server` roles,

> so that requests with `$JUPYTERHUB_API_TOKEN` will have permission to read the access token:

Reviewer: Is there likely to be anything in `auth_state` that an admin wouldn't want a user to see?

Author: It's worth mentioning this, because it is possible in theory and there are not currently more fine-grained controls. I don't believe so with the current OAuthenticators, at least, which have the userinfo and token responses. Essentially, all the information OAuthenticator puts in auth state is retrievable with the access token we are after, so if you are doing this to pass the access token, access to auth state isn't giving access to any more info than you are already meaning to give by relaying the access token itself.

Author: Added a caveat to the top of this section

    {
        "name": "user",
        "scopes": [
            "self",
            "admin:auth_state!user",
        ],
    },
    {
        "name": "server",
        "scopes": [
            "users:activity!user",
            "access:servers!server",
            "admin:auth_state!user",
        ],
    },
]
```

We then also need to make sure "auth state" is enabled
(it is enabled by default in the jupyterhub helm chart):

```python
c.Authenticator.enable_auth_state = True
# also set $JUPYTERHUB_CRYPT_KEY env to 32-byte string
# e.g. with `openssl rand -hex 32`
```

At this point:

1. When a user logs in, the OAuth user info and access token are encrypted and persisted in the Hub database.
2. When the server token requests the user model at `/hub/api/user`, an `auth_state` field will be present, containing the current auth state.
3. Further, when accessing `/hub/api/user` the `refresh_user` logic is triggered if `auth_refresh_age` has elapsed since the last refresh.

This means that you can access `/hub/api/user` with `$JUPYTERHUB_API_TOKEN` and it will **always return a valid access token**,
even if the currently stored token has expired when the request is made.

> To retrieve the access token, make a request to `${JUPYTERHUB_API_URL}/hub/user` with `${JUPYTERHUB_API_TOKEN}, e.g. from Python:

Reviewer: ```suggestion
To retrieve the access token, make a request to `${JUPYTERHUB_API_URL}/hub/user` with `${JUPYTERHUB_API_TOKEN}`, e.g. from Python:
```

  a new access token is retrieved via the [refresh_token grant](https://www.oauth.com/oauth2-servers/making-authenticated-requests/refreshing-an-access-token/)
- If no auth info is retrievable (e.g. no refresh token and access token is expired or both are expired or revoked),
  then the user must login again before they are able to take actions in JupyterHub
  because at this point their authorization state is unknown and could no longer be valid.

There is also an option [Authenticator.refresh_pre_spawn](inv:jupyterhub:py:attribute#jupyterhub.auth.Authenticator.refresh_pre_spawn) which can be enabled:

```python
c.Authenticator.refresh_pre_spawn = True
```

to ensure auth is up-to-date before launching a server.
This is most useful when the server is being passed an access token
because it ensures the token is valid when the server starts.

## Refreshing tokens from user sessions

```{warning}
This example requires granting users read access to their own `auth_state`.
If you plan to provide users with access tokens,
`auth_state` does not typically include information your users won't have access to with the token itself,

> but it is worth making sure that your Authenticator configuration places anything in `auth_state`

Reviewer: ```suggestion
but it is worth making sure that your Authenticator configuration doesn't put anything in `auth_state`
```
Go ahead and merge after this fix (or equivalent)

Author: Thanks!
